From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 8 Mar 2025 17:21:16 +0100
Subject: [PATCH 086/225] net: wwan: add simple rpmsg-based network driver
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The rpmsg framework is often used for communicating with a built-in
modem processor in various mobile SoCs. On many Unisoc SoCs with LTE
modems, rpmsg channels are available that can be used directly for
transmitting and receiving raw IP packets. Add a simple network driver
based on the rpmsg API for such modems.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/net/wwan/Kconfig          |  11 +++
 drivers/net/wwan/Makefile         |   1 +
 drivers/net/wwan/rpmsg_wwan_net.c | 132 ++++++++++++++++++++++++++++++
 3 files changed, 144 insertions(+)
 create mode 100644 drivers/net/wwan/rpmsg_wwan_net.c

diff --git a/drivers/net/wwan/Kconfig b/drivers/net/wwan/Kconfig
index 88df55d78..ea1c9c855 100644
--- a/drivers/net/wwan/Kconfig
+++ b/drivers/net/wwan/Kconfig
@@ -93,6 +93,17 @@ config RPMSG_WWAN_CTRL
 	  To compile this driver as a module, choose M here: the module will be
 	  called rpmsg_wwan_ctrl.
 
+config RPMSG_WWAN_NET
+	tristate "RPMSG WWAN network driver"
+	depends on RPMSG
+	help
+	  This driver provides access to raw IP network data channels available
+	  via RPMSG. This is mainly used for modems integrated into Unisoc
+	  SoCs, which provide SIPC channels for mobile data.
+
+	  To compile this driver as a module, choose M here: the module will be
+	  called rpmsg_wwan_net.
+
 config IOSM
 	tristate "IOSM Driver for Intel M.2 WWAN Device"
 	depends on PCI
diff --git a/drivers/net/wwan/Makefile b/drivers/net/wwan/Makefile
index 3960c0ae2..2e065f709 100644
--- a/drivers/net/wwan/Makefile
+++ b/drivers/net/wwan/Makefile
@@ -12,5 +12,6 @@ obj-$(CONFIG_MHI_WWAN_CTRL) += mhi_wwan_ctrl.o
 obj-$(CONFIG_MHI_WWAN_MBIM) += mhi_wwan_mbim.o
 obj-$(CONFIG_QCOM_BAM_DMUX) += qcom_bam_dmux.o
 obj-$(CONFIG_RPMSG_WWAN_CTRL) += rpmsg_wwan_ctrl.o
+obj-$(CONFIG_RPMSG_WWAN_NET) += rpmsg_wwan_net.o
 obj-$(CONFIG_IOSM) += iosm/
 obj-$(CONFIG_MTK_T7XX) += t7xx/
diff --git a/drivers/net/wwan/rpmsg_wwan_net.c b/drivers/net/wwan/rpmsg_wwan_net.c
new file mode 100644
index 000000000..1b18ca69b
--- /dev/null
+++ b/drivers/net/wwan/rpmsg_wwan_net.c
@@ -0,0 +1,132 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copyright (C) 2025 Otto Pflüger
+ */
+
+#include <linux/etherdevice.h>
+#include <linux/if_arp.h>
+#include <linux/rpmsg.h>
+#include <net/pkt_sched.h>
+
+struct rpmsg_wwan_net {
+	struct rpmsg_endpoint *ept;
+};
+
+static int rpmsg_wwan_net_callback(struct rpmsg_device *rpdev, void *buf,
+				   int len, void *priv, u32 addr)
+{
+	struct net_device *ndev = dev_get_drvdata(&rpdev->dev);
+	struct sk_buff *skb;
+
+	skb = netdev_alloc_skb(ndev, len);
+	if (!skb)
+		return -ENOMEM;
+
+	skb_put_data(skb, buf, len);
+
+	switch (skb->data[0] & 0xf0) {
+	case 0x40:
+		skb->protocol = htons(ETH_P_IP);
+		break;
+	case 0x60:
+		skb->protocol = htons(ETH_P_IPV6);
+		break;
+	default:
+		netdev_warn(ndev, "unexpected protocol, first byte %02x\n",
+			    skb->data[0]);
+		return 0;
+	}
+
+	netif_rx(skb);
+
+	return 0;
+}
+
+static netdev_tx_t rpmsg_wwan_net_start_xmit(struct sk_buff *skb,
+					     struct net_device *ndev)
+{
+	struct rpmsg_wwan_net *rpwwan = netdev_priv(ndev);
+	int ret;
+
+	ret = rpmsg_trysend(rpwwan->ept, skb->data, skb->len);
+	if (ret) {
+		netdev_err(ndev, "rpmsg tx failed: %d\n", ret);
+		return NETDEV_TX_BUSY;
+	}
+
+	return NETDEV_TX_OK;
+}
+
+static const struct net_device_ops rpmsg_wwan_net_ops = {
+	.ndo_start_xmit = rpmsg_wwan_net_start_xmit,
+};
+
+static void rpmsg_wwan_netdev_setup(struct net_device *ndev)
+{
+	ndev->netdev_ops = &rpmsg_wwan_net_ops;
+	ndev->type = ARPHRD_RAWIP;
+	ndev->flags = IFF_POINTOPOINT | IFF_NOARP;
+	ndev->mtu = ETH_DATA_LEN;
+	ndev->max_mtu = ETH_DATA_LEN;
+	ndev->tx_queue_len = DEFAULT_TX_QUEUE_LEN;
+
+	ndev->addr_assign_type = NET_ADDR_RANDOM;
+	eth_random_addr(ndev->perm_addr);
+}
+
+static const struct device_type wwan_type = { .name = "wwan" };
+
+static int rpmsg_wwan_net_probe(struct rpmsg_device *rpdev)
+{
+	struct rpmsg_wwan_net *rpwwan;
+	struct net_device *ndev;
+	int ret;
+
+	ndev = alloc_netdev(sizeof(*rpwwan), "wwan%d", NET_NAME_ENUM,
+			    rpmsg_wwan_netdev_setup);
+	if (!ndev)
+		return -ENOMEM;
+
+	SET_NETDEV_DEV(ndev, &rpdev->dev);
+	SET_NETDEV_DEVTYPE(ndev, &wwan_type);
+
+	rpwwan = netdev_priv(ndev);
+	rpwwan->ept = rpdev->ept;
+
+	dev_set_drvdata(&rpdev->dev, ndev);
+
+	ret = register_netdev(ndev);
+	if (ret) {
+		free_netdev(ndev);
+		return ret;
+	}
+
+	return 0;
+}
+
+static void rpmsg_wwan_net_remove(struct rpmsg_device *rpdev)
+{
+	struct net_device *ndev = dev_get_drvdata(&rpdev->dev);
+
+	unregister_netdev(ndev);
+}
+
+static const struct of_device_id rpmsg_wwan_net_of_table[] = {
+	/* RPMSG channels for Unisoc SoCs with integrated LTE modem */
+	{ .compatible = "sprd,seth" },
+	{ }
+};
+MODULE_DEVICE_TABLE(of, rpmsg_wwan_net_of_table);
+
+static struct rpmsg_driver rpmsg_wwan_net_driver = {
+	.drv.name = "rpmsg-wwan-net",
+	.drv.of_match_table = rpmsg_wwan_net_of_table,
+	.probe = rpmsg_wwan_net_probe,
+	.remove = rpmsg_wwan_net_remove,
+	.callback = rpmsg_wwan_net_callback,
+};
+module_rpmsg_driver(rpmsg_wwan_net_driver);
+
+MODULE_LICENSE("GPL");
+MODULE_DESCRIPTION("Simple RPMSG-based WWAN network driver");
+MODULE_AUTHOR("Otto Pflüger <otto.pflueger@abscue.de>");
